rubyforgood/casa
# [780] Allow admin to edit admin users

### What github issue is this PR for, if any?

Resolves #780 

### What changed, and why?

1. Admins can now edit the `email` and `display_name` of other casa admin users;
2. A new resource was introduced `/casa_admins`, for routes `edit` and `update`;
3. A few `request`, `view`, and `feature` specs were introduced to cover the new code execution path.

### How will this affect user permissions?

- Volunteer permissions: Can't access the edit form.
- Supervisor permissions: Can't access the edit form.
- Admin permissions: Only admins will have access to the edit form.

### How is this tested? (please write tests!) 💖💪

For manual tests, one should:

Happy path:

1. Log in as an admin user;
2. On the dashboard page, scroll down to the `Admins` section;
3. Click on the **new** `Edit` link next to a casa admin user listed in that section;
4. Update the user with new information and click `Update`;
5. Verify that you are redirected back to the dashboard page, with a success message on the top of the page.

Specs were added to cover a few scenarios including validations:

**Scenario 1**

```
Given that I am logged into CASA as an admin user
And I am on the dashboard page
I should be able to click on the `Edit` link next to another admin user listed under the `Admins` section
And be taken to an edit admin user form
```

**Scenario 2**

```
Given that I am logged into CASA as an admin user
And I am on the edit admin form page
I should be able to edit the admin's email and display name but not their password
```

**Scenario 3**

```
Given that I am logged into CASA as an admin user
And I am on the edit admin form page
When attempting to update the user without an email or display name
Then I should see validation error messages rendered on the form along with the visual indicators on the form fields.
```

### Screenshots please :)

![2020-09-22 10 04 17](https://user-images.githubusercontent.com/508128/93887054-90b42e00-fcbc-11ea-80d9-687f0d3da2d5.gif)

### Feelings gif (optional)

![success](https://media.giphy.com/media/7kO9VZjCv3FkI/giphy.gif)
